#4: Added requirments document (#13)
exasol/advanced-analytics-framework: doc/system_requirements.md

# System Requirement Specification Advanced Analytics Framework

## Introduction

> This library is an event-driven framework that allows users to build complex 

Reviewer: event driven is already part of the design

data processing algorithms in Exasol.

## About This Document

### Goal

The goal of this library is to provide a general framework to implement complex 
data analysis algorithms with Exasol. This framework provides certain features 
to users in which they are able to run their implementations.

## Stakeholders

Software developers use this framework to develop complex data processing 
algorithms such as machine learning methods.

## Features

Features are the highest level requirements in this document that describe the 
main functionality of Advanced Analytics Framework.

> 

Reviewer: The algorithm is implemented by the user, so the interface needs to be simple for everyone to implement algorithms

> ### Event Loop

Reviewer: that is already design

`feat~event-loop~1`

Event Loop processes only the state transitions by executing queries returned by 
the Event Handler. Some complex algorithms need a global loop. In each iteration 
of the loop they need to execute SQL queries until the algorithm no longer 
needs to run further queries. 

Needs: req

### Event Handler
`feat~event-handler~1`

TODO

Needs: req

## Functional Requirements

This section lists functional requirements. The requirements are grouped by 
feature where they belong to a single feature.

> 

Reviewer: We want to implement the algorithm in python, because it has a larger community and better tooling and other libs are already in this language

> 

Reviewer: Algorithm results from a query or return a value (json value)

> ### Event Loop

Reviewer: Global Loop

#### Calling Event-Handler
`req~calling-event-handler~1`

It takes the configuration of an operation, which is implemented by user, as 
input. It builds query using this configuration to call Event Handler. 
Furthermore, in each iteration the return query of the Event Handler is called 
again as the next state action and performed the new state transition. 

Covers:

* [feat~event-loop~1](#event-loop)

Needs: dsn

#### Running Queries
`req~running-queries~1`

The output of the called Event Handler operation contains the necessary SQLs 
if the algorithm needs to continue. This function runs these required SQLs.

Covers:

* [feat~event-loop~1](#event-loop)

Needs: dsn

#### Generating Return Query
`req~generating-return-query~1`

When the operation needs to continue, the Event Handler returns the next query 
of the Event-Handler that should get called again.  This method builds the 
return query using the current state information and make it ready to be 
called as a new state.

Covers:

* [feat~event-loop~1](#event-loop)

Needs: dsn

### Event Handler

#### Handling States
`req~TODO~1`

Covers:

* [feat~event-handler~1](#event-handler)

Needs: dsn

> #### Managing Storage

Reviewer: Managing Temporary BucketFS Files

Reviewer: All bucketfs files that will be kept as result, should be placed in the same directory in the bucketfs

`req~TODO~1`

Covers:

* [feat~event-handler~1](#event-handler)

Needs: dsn

#### Managing Temporary Tables
`req~TODO~1`

Covers:

* [feat~event-handler~1](#event-handler)

> Needs: dsn

Reviewer: Error Handling:
- Cleanup Temporary Tables, Temporary BucketFS Files
- User can set option keep non-temporary things, for debugging keeping it could be useful or for checkpointing

This document describes the system requirements for the advanced analytics 
framework on which complex algorithms can be built.

## About This Document

### Goal

The goal of this library is to provide a general framework to implement complex 
data analysis algorithms with Exasol. This framework provides certain features 
to users in which they are able to run their implementations.

## Stakeholders

Software developers use this framework to develop complex data processing 
algorithms such as machine learning methods.

## Features

Features are the highest level requirements in this document that describe the 
main functionality of Advanced Analytics Framework.

### Global Loop
`feat~global-loop~1`

Some complex algorithms need a global loop. In each iteration of the loop, they 
need to execute SQL queries until the algorithm no longer needs to run 
further queries. 

Needs: req

### Implementation Framework
`feat~implementation-framework~1`

The logic of the algorithm is implemented by user. The implementation needs 
to call a framework which provides certain features to the user code, such as 
automatic state handling. 

Needs: req

### Error Handling
`feat~error-handling~1`

> Any errors, that may occur during state transitions or  the execution of queries, 

Reviewer: ```suggestion
Any errors, that may occur during user code execution or the execution of queries, 
```

should be handled. Temporarily created files and tables need to be cleaned.
User can set option to keep temporary records. It can be useful to keep these 
for debugging.

Needs: req

## Functional Requirements

This section lists functional requirements. The requirements are grouped by 
feature where they belong to a single feature.

### Global Loop

#### Initiating Global Loop
`req~intiating-global-loop~1`

> It expects the configuration of the operation, which is implemented by user, as 

Reviewer: Who is it here

> input. It needs to build starting query using this configuration. 

Reviewer: ```suggestion
input. It needs to build starting queries using this configuration. 
```

Covers:

* [feat~global-loop~1](#global-loop)

Needs: dsn

#### Iterating over Loop
`req~iterating-over-loop~1`

> At each iteration, the framework checks the output of the called operation query 

Reviewer: ```suggestion
At each iteration, the framework asks the user code to check the output of the executed query 
```

> to decide whether to continue. If the output contains the necessary SQLs and the 
> status information indicating that it is in progress, these SQLs should run as 
> the next state action and perform the new state transition. 

Reviewer: ```suggestion
to decide whether to continue. If the user code decides to stop the framework returns the results returned by the user code. Otherwise, the user code can start another iteration with further queries.
```

Covers:

* [feat~global-loop~1](#global-loop)

Needs: dsn

#### Generating Return Query
`req~generating-return-query~1`

When the operation needs to continue, the output of the operation query returns
the next query that should get called again.  It is required to build the  
return query using the current state information and make it ready to be  

> called as a new state.

Reviewer: Still too much design

Needs: dsn

#### Managing Temporary BucketFS Files
`req~managing-temporary-bucketfs-files~1`

The algorithm should be able to access BucketFS and create temporary BucketFS 
files that will be kept as result. These temporary file should be placed  in the 
same directory in the BucketFS so that it will be easy to access them.

Covers:

* [feat~implementation-framework~1](#implementation-framework)

Needs: dsn

#### Managing Temporary Tables
`req~managing-temporary-tables~1`

The framework uses temporary tables to store large intermediate results. The 
system should make sure that these are cleaned up.

Covers:

* [feat~implementation-framework~1](#implementation-framework)

Needs: dsn

#### Message Passing in Distributed Computation
`req~message-passing-in-distributed-computation~1`

The data used by the algorithm might be distributed to the UDF instances. The 
framework should enable message passing between these UDF instances to access 
data or intermediate results of the other partitions.

Covers:

* [feat~implementation-framework~1](#implementation-framework)

Needs: dsn

#### Electing Leader
`req~electing-leader~1`

> The framework needs to elect a leader from among the UDF instances.

Reviewer: ```suggestion
The message passing framework needs to be able to elect a leader from among the UDF instances.
```

Covers:

* [feat~implementation-framework~1](#implementation-framework)

Needs: dsn

#### Achieving Parallelism
`req~collective-operation~1`

> The framework must support Collective operation on UDF instances to achieve 

Reviewer: ```suggestion
The message passing framework must support Collective operation on UDF instances to achieve 
```

Features are the highest level requirements in this document that describe the 
main functionality of Advanced Analytics Framework.

### Global Loop
`feat~global-loop~1`

Some complex algorithms need a global loop. In each iteration of the loop, they 
need to execute SQL queries until the algorithm no longer needs to run 
further queries. 

Needs: req

### Implementation Framework
`feat~implementation-framework~1`

The logic of the algorithm is implemented by user. The implementation needs 
to call a framework which provides certain features to the user code, such as 

> automatic state handling. 

Reviewer: Here it is better to use things like temporary table handling and running the queries and returning the results instead of the automatic state handling, because the state handling is only necessary in our design. While temporary table handling and running the queries and returning the results, are the things the framework should take care of for the user and which we potentially want replace with different implementations.
